### PR TITLE
Add refined debouncer

### PR DESCRIPTION
**Pull request by dfaust**
This is another take at event debouncing.
Unlike the `mini` debouncer, the `refined` debouncer is much more opinionated.
It tries its best to simplify event handling and in particular can use file system ids to stitch together rename events in case the notification back-end doesn't support cookies.

Features:
* Only emits a single `Rename` event if the rename `From` and `To` events can be matched
* Merges multiple `Rename` events
* Optionally keeps track of the file system IDs all files and stitches rename events together (FSevents, Windows)
* Emits only one `Remove` event when deleting a directory (inotify)
* Doesn't emit duplicate create events
* Doesn't emit `Modify` events after a `Create` event

This PR also removes the `rename` thread from the inotify back-end and simplifies the rename event handling.
Because of that I bumped the version number to 6.0.0.

I haven't done much testing, yet. But it works well on Linux so far.

This PR also implements `Copy` for `EventKind` and `ModifyKind` (#458).
~~And I changed the dependency declarations to use workspace dependencies.~~ (Requires Cargo 1.64)

### CODE REVIEW

**Comment by 0xpr03** [comment]
I also want to throw another idea at you: I was thinking about whether we could make a "fat" debouncer that also detects filesystems that require usage of pollwatcher. It could then automatically use the pollwatcher backend in addition to the native one. (See #475 ) This would of course mean detecting the added path in the mount section for linux. It would already help with missing events for WSL environments that watch windows host folders

**Comment by JohnTitor** [comment]
Meanwhile, could you split https://github.com/notify-rs/notify/issues/458 as a separate PR? It's essentially unrelated to the debouncer change here and it's worth releasing it as v5.2.0.

**Comment by dfaust** [comment]
> I was thinking about whether we could make a "fat" debouncer ...

@0xpr03 
I like the name "fat" debouncer. "refined" was the best name I could come up with.
I'm not sure how combining the poll watcher with the native watcher would look like,
but there are a few areas where a debouncer and the poll watcher can be integrated closer.
I was thinking about sharing the file ID cache (this is also a reason why I made the `FileIdCache` trait).
Going even further it may be possible to create some sort of "offline" change detection, where events are reported for changes that happened while the application wasn't running.

> could you split #458 as a separate PR?

@JohnTitor 
Sure.

File: Cargo.toml
Comment (0xpr03) [inline Cargo.toml:6]: let's change this to "notify-debouncer-full"

File: notify-debouncer-refined/src/lib.rs
Comment (0xpr03) [inline notify-debouncer-refined/src/lib.rs:582]: if you want to, you can already move ahead of #467 and use a tick-less system - otherwise we could introduce this later on

the idea is to either let a new event wake up the debouncer, or schedule a sleep-timeout if we need another tick (remaining events in the event queue)

I'll merge #467 after this one for the mini debouncer, and can also do the followup for this one, if you want

File: notify-debouncer-refined/Cargo.toml
Comment (0xpr03) [inline notify-debouncer-refined/Cargo.toml:23]: I think we can just break the version compatibility here, if we're doing a major release anyway for notify itself

So feel free to just undo the 1.56 compatibility thingy and I may clean up stuff like this later on

File: CHANGELOG.md
Comment (0xpr03) [inline CHANGELOG.md:7]: can confirm this

**Review by 0xpr03 — CHANGES_REQUESTED** [review]
Looks overall pretty good. A ton more logic in the debouncer now, but this will give a much better user experience.

As noted I would suggest renaming to -debouncer-full.
Fat is also an option, but sounds more negative than it has to.

I would drop 1.56 compatibility and move on, as for the crossbeam things we could really need some functionality from 1.60+

I really like the testing harness.

File: file-id/Cargo.toml
Comment (0xpr03) [inline file-id/Cargo.toml:1]: just to confirm, we don't have to publish this, right ?

File: notify-debouncer-refined/src/lib.rs
Comment (dfaust) [reply notify-debouncer-refined/src/lib.rs]: I only saw #467 after I finished my PR. I think we should merge it first and then improve it.

File: notify-debouncer-refined/Cargo.toml
Comment (dfaust) [reply notify-debouncer-refined/Cargo.toml]: I undid the 1.56 compatibility commit and raised the MSRV to 1.60.

File: file-id/Cargo.toml
Comment (dfaust) [reply file-id/Cargo.toml:1]: That's right.
I think there could be some utility in publishing it, but there is also room for improvement.
Currently it only supports the old, low resolution, ids on windows. Newer file systems have file ids with a size 128 bits instead of 64 bits.

File: examples/Cargo.toml
Comment (dfaust) [reply examples/Cargo.toml:9]: In the long term it would be nice to use workspace dependencies. That would make things simpler. But they require Rust 1.64.

**Comment by 0xpr03** [comment]
> I think we should merge it first and then improve it.

If you have something you want to merge pre 6.0 release leave me comment, otherwise I'd publish a v6 tomorrow.

**Comment by dfaust** [comment]
@0xpr03 
Will you update the debouncer-full as well, after merging #467, or should I do that? Please let me know.

I think a new release is great, but as I said, I haven't done much testing, yet.
Maybe we could include a call for testing in the release notes.

**Comment by 0xpr03** [comment]
Yeah I'll try to do that as well.

> call for testing 

We could do a RC

**Comment by dfaust** [comment]
> We could do a RC (but this time not for 2 years)

:smile: I'll leave that decision to you.
Personally, since this is the first release of the full debouncer, I would release it as version 0.1.0 and just add a few lines to the release notes saying something like: "This release contains a new debouncer. It should still be considered experimental at this point. Please let us know if you encounter any issues with it."

**Comment by 0xpr03** [comment]
Ah yeah I meant RC for v6 of notify. But I guess we'll skip that. And for the debouncer-full a 0.1.0 is fine.

**Comment by 0xpr03** [comment]
This got a little bit hairy, apart from me not noticing that the updated versions weren't all correct, crates.io does require publishing file-id. So that's now also up there.

**Pull request merged**
